File: midolman/src/main/java/org/midonet/cluster/data/TraceRequest.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.cluster.data;

import org.midonet.midolman.rules.Condition;

import java.util.Objects;
import java.util.UUID;

public class TraceRequest
    extends Entity.Base<UUID, TraceRequest.Data, TraceRequest> {

    public TraceRequest() {
        super(null, new Data());
    }

    public String getName() {
        return getData().name;
    }

    public TraceRequest setName(String name) {
        getData().name = name;
        return this;
    }

    public UUID getDeviceId() {
        return getData().deviceId;
    }

    public TraceRequest setDeviceId(UUID deviceId) {
        getData().deviceId = deviceId;
        return this;
    }

    public DeviceType getDeviceType() {
        return getData().deviceType;
    }

    public TraceRequest setDeviceType(DeviceType deviceType) {
        getData().deviceType = deviceType;
        return this;
    }

    public Condition getCondition() {
        return getData().condition;
    }

    public TraceRequest setCondition(Condition condition) {
        getData().condition = condition;
        return this;
    }

    public long getLimit() {
        return getData().limit;
    }

    public TraceRequest setLimit(long limit) {
        getData().limit = limit;
        return this;
    }

    public UUID getEnabledRule() {
        return getData().enabledRule;
    }

    public TraceRequest setEnabledRule(UUID enabledRule) {
        getData().enabledRule = enabledRule;
        return this;
    }

    @Override
    public TraceRequest self() {
        return this;
    }

    public enum DeviceType {
        BRIDGE, PORT, ROUTER
    }

    public static class Data  {
        public String name;
        public DeviceType deviceType;
        public UUID deviceId;
        public Condition condition;
        public long limit;
        public UUID enabledRule;

        @Override
        public boolean equals(Object other) {
            if (this == other) {
                return true;
            }
            if (other == null || !(other instanceof Data)) {
                return false;
            }

            System.out.println("other " + other + " this " + this);
            Data that = (Data)other;
            return Objects.equals(name, that.name)
                && deviceType == that.deviceType
                && Objects.equals(deviceId, that.deviceId)
                && Objects.equals(condition, that.condition)
                && limit == that.limit
                && Objects.equals(enabledRule, that.enabledRule);
        }

        @Override
        public int hashCode() {
            return Objects.hash(name, deviceType, deviceId,
                                condition, limit, enabledRule);
        }

        @Override
        public String toString() {
            return "TraceRequest.Data{name=" + name
                + ", deviceType=" + deviceType
                + ", deviceId=" + deviceId
                + ", condition=" + condition
                + ", limit=" + limit
                + ", enabledRule=" + enabledRule + "}";
        }
    }
}

File: tests/mdts/lib/tracerequest.py
# Copyright 2015 Midokura SARL
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from mdts.lib.resource_base import ResourceBase

from rule import FIELDS

# subset of condition fields (those that are easier to handle)
CONDFIELDS = [
    'dl_src',
    'inv_dl_src',
    'dl_src_mask',
    'dl_dst',
    'inv_dl_dst',
    'dl_dst_mask',
    'dl_type',
    'inv_dl_type',
    'nw_src_address',
    'nw_src_length',
    'inv_nw_src',
    'nw_dst_address',
    'nw_dst_length',
    'inv_nw_dst',
    'nw_proto',
    'inv_nw_proto',
    'nw_tos',
    'inv_nw_tos',
    'tp_src',
    'inv_tp_src',
    'tp_dst',
    'inv_tp_dst',
    'match_forward_flow',
    'match_return_flow',
    'fragment_policy',
]


class TraceRequest(ResourceBase):
    def __init__(self, api, context, data):
        super(TraceRequest, self).__init__(api, context, data)

    def build(self):
        self._mn_resource = self._api.add_tracerequest()
        self._mn_resource.set_name(self._get_name())
        if self._data.has_key('limit'):
            self._mn_resource.set_limit(self._data['limit'])
        if self._data.has_key('enabled'):
            self._mn_resource.set_enabled(self._data['enabled'])
        else:
            self._mn_resource.set_enabled(False)

        if self._data.has_key('port'):
            port = self._data['port']
            portdev = self._context.get_device_port(
                port['device'], port['portid']).get_mn_resource()
            portid = portdev.get_id()
            self._mn_resource.set_port(portid)
        elif self._data.has_key('router'):
            router = self._context.get_router(
                self._data['router']).get_mn_resource()
            self._mn_resource.set_router(router.get_id())
        elif self._data.has_key('bridge'):
            bridge = self._context.get_bridge(
                self._data['bridge']).get_mn_resource()
            self._mn_resource.set_bridge(bridge.get_id())

        for f in CONDFIELDS:
            if self._data.has_key(f):
                getattr(self._mn_resource, f)(self._data[f])
        self._mn_resource.create()

    def set_enabled(self, enabled):
        self._mn_resource.set_enabled(enabled)
        self._mn_resource.update()

    def destroy(self):
        self._mn_resource.delete()

    def get_id(self):
        return self._mn_resource.get_id()
